x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Limit.java

```diff
@@ -21,21 +21,37 @@ public class Limit extends UnaryPlan {
     public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(LogicalPlan.class, "Limit", Limit::new);
 
     private final Expression limit;
-
-    public Limit(Source source, Expression limit, LogicalPlan child) {
+    /**
+     * Important for optimizations. This should be {@code true} in most cases, which allows this instance to be duplicated past a child plan
+     * node that increases the number of rows, like for LOOKUP JOIN and MV_EXPAND.
+     * Needs to be set to {@code false} in {@link org.elasticsearch.xpack.esql.optimizer.rules.logical.PushDownAndCombineLimits} to avoid
+     * infinite loops from adding a duplicate of the limit past the child over and over again.
+     */
+    private final boolean allowDuplicatePastExpandingNode;
+
+    public Limit(Source source, Expression limit, LogicalPlan child, boolean allowDuplicatePastExpandingNode) {
         super(source, child);
         this.limit = limit;
+        this.allowDuplicatePastExpandingNode = allowDuplicatePastExpandingNode;
     }
 
     private Limit(StreamInput in) throws IOException {
-        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(LogicalPlan.class));
+        this(
+            Source.readFrom((PlanStreamInput) in),
+            in.readNamedWriteable(Expression.class),
+            in.readNamedWriteable(LogicalPlan.class),
+            true
+        );
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         Source.EMPTY.writeTo(out);
         out.writeNamedWriteable(limit());
         out.writeNamedWriteable(child());
+        // For limits sent to data nodes, allowDuplicatePastExpandingNode should always be true.
```

Comment:
Is this a real constraint? If it is, then we could have an `assert` here to be sure.
On the other hand, this is a transient setting: even if a data node receives a `false`, next logical planner iteration will switch it to `true` with no real consequences (apart from one more optimization loop). If we want to go this way, we could make it clear in the comment.

Reply:
I've seen that `MvExpand` actually asserted that. It's less important here because in case of `MvExpand`, not serializing the limit is incorrect, while not serializing `allowDuplicatePastExpandingNode` at most leads to another copy of the limit being sent down as you described.

But it's worth noticing in tests when this behavior changes, so I'll add the assertion back in + add a comment!